[dataquery] test plan (#9691)
aces/Loris: modules/dataquery/test/TestPlan.md

## Table of Contents

1. [Setup](#setup)
2. [Panels](#panels)
3. [Recent Queries](#recent-queries)
4. [Pin Tests](#pin-tests)
5. [Fields Selection](#fields-selection)
6. [Other parameters](#other-parameters)
7. [Sync not Sync](#sync-not-sync)
8. [Filters](#filters)
9. [Run Query Page](#run-query-page)

## Setup

Sign into your loris instance with an **admin** account.

- select Admin Accounts and click the Add User button.
- enter all required information.
- select *ONE* site for example, "Rome".
- select *ONE* project, for example Pumpernickel.
- [x] Access Profile: View/Create Candidates and Timepoints - Own Sites.
- [x] Data Dictionary: View Parameter Type Descriptions
- [x] Data Query Tool: View/Download Cross-Modality Data

## Panels

- Ensure the module loads only for a user that has the `dataquery_view` permission. They must also have access to the dictionary module.

- Assert that the following panels collapse and open

   1. Instructions
   2. Recent Queries
   3. Next Steps (bottom-right corner)

- Assert that `Continue to Define Fields` button in the main panel, and `Choose Fields` button in the `Next Steps` panel are redirecting to the same page.

## Recent Queries

> ### Perform the following assertions: (If no queries are available, add some)

Reviewer: I will try something here, so bare with me. I had some difficulties reading the document, and I think it is linked to the document format. I think we should have a consistent difference between two types of lines in this document: 
1. `assertion/checks to do`, right now that is ok, they are listed.
2. `instructions-like lines`, such as "Repeat for `Shared Only`, `Named Only` and `No run times` checkboxes." or "Make a mix of the following actions" have varying formats.

So applying the markdown quote (it can also be multi-line) could do the trick for instruction like?
```suggestion
> Perform the following assertions: (If no queries are available, add some)
```

- Assert that the queries you made have their parameters correctly displayed (i.e. fields and filters).

- Assert that `text filter` immediately filter the queries.

- Assert that clicking `Collapse queries` effectively collapses all queries.

> - Make a mix of the following actions
> 
>    1. `Star`: should have a yellow star
>    2. `Name` some queries
>    3. `Share` should have a blue shared state
>    4. `Rerun` some queries

Reviewer: ```suggestion
> Make a mix of the following actions
>
>   1. `Star`: should have a yellow star
>   2. `Name` some queries
>   3. `Share` should have a blue shared state
>   4. `Rerun` some queries
```

- Assert that using the `Starred Only` checkbox filter only keeps the starred ones.

- Assert that removing the `Starred Only` checkbox prints the same queries originally printed.

> Repeat for `Shared Only`, `Named Only` and `No run times` checkboxes.

Reviewer: ```suggestion
> Repeat for `Shared Only`, `Named Only` and `No run times` checkboxes.
```

- Assert that mixing checkboxes returns the right query selection. Only queries that match all of the checked conditions should be displayed.

- Remove all pinned queries and assert that there is no `Study Queries` panel at the top of the page.

> ### If all of these actions work, then this test is successful

Reviewer: True, to be more precise, if any line fails, the line should be reported
```suggestion
```

## Pin Tests

> Click the `Pin` icon to perform the following tests (ensure that you have 'admin' permission):

Reviewer: ```suggestion
> Click the `Pin` icon to perform the following tests (ensure that you have 'admin' permission):
```

> - With the `query name` text field empty, click the `Submit` button and assert that the error message reads:
> 
> >`Must provide a query name to pin query as.'

Reviewer: ```suggestion
- With the `query name` text field empty, click the `Submit` button and assert that the error message reads: `Must provide a query name to pin query as.'
```

> - Enter a query name

Reviewer: ```suggestion
> Enter a query name.
```

> - Uncheck the following:
> - [ ] `Pin Study Query`
> - [ ] `Pin Dashboard Summary`
> - [ ] `Pin to Login Page`

Reviewer: ```suggestion
> Uncheck the following:
>  - [ ] `Pin Study Query`
>  - [ ] `Pin Dashboard Summary`
>  - [ ] `Pin to Login Page`
```

# Data Query - Test Plan

This test plan is long, so, it is divided into sections. As such, each test in the table of contents is meant to be run with 2 users; one with one site and one project, the other with multiple sites and multiple projects.

## Table of Contents

1. [Permission Test](#permission-test)
2. [Panels Test](#panels-test)
3. [Recent Queries Test](#recent-queries-test)
4. [Pin Tests](#pin-tests)
5. [Fields Selection](#fields-selection)
6. [Other parameters](#other-parameters)
7. [Sync not Sync](#sync-not-sync)
8. [Filters](#filters)
9. [Run Query Page](#run-query-page)

- Create a user with one site and one project, for example: Rome, Pumpernickel.
- Create a second user with more than one site and more than one project with the same permissions.

## Permission Test

For your first user, enable the following permissions:
- [x] Access Profile: View/Create Candidates and Timepoints - Own Sites. `data_entry`  
- [x] Data Dictionary: View Parameter Type Descriptions. `data_dict_edit`
- [x] Data Query Tool: View/Download Cross-Modality Data. `dataquery_view`

> - Remove the following permission and assert that the module does **not** load:
> - [x] Data Query Tool: View/Download Cross-Modality Data. `dataquery_view`
> - Switch to your other user, remove the `data_query_view` permission and assert that the module does **not** load for this user
> 

Reviewer 2: ```suggestion
Remove the following permission and assert that the module does **not** load:
- [x] Data Query Tool: View/Download Cross-Modality Data. `dataquery_view`
Switch to your other user, remove the `data_query_view` permission and assert that the module does **not** load for this user

```

## Panels Test

1. Assert that the following panels collapse and open :

   - Instructions
   - Recent Queries
   - Next Steps (in the bottom-right corner)

2. Assert that `Continue to Define Fields` button in the main panel, and `Choose Fields` button in the `Next Steps` panel are redirecting to the same page.

## Recent Queries Test

1. Assert that the queries you made have their parameters correctly displayed (i.e. fields and filters).
2. Assert that `text filter` immediately filters the queries.
3. Assert that clicking `Collapse queries` effectively collapses all queries.

Make a mix of the following actions

- `Star`: should have a yellow star
- `Name` some queries
- `Share` should have a blue shared state
- `Rerun` some queries

> 3. Assert that using the `Starred Only` checkbox filter only keeps the starred ones.
> 4. Assert that removing the `Starred Only` checkbox prints the same queries originally printed.
> 
> - Repeat 3. and 4. for `Shared Only`, `Named Only` and `No run times` checkboxes.
> 
> 1. Assert that mixing checkboxes returns the right query selection. Only queries that match all of the checked conditions should be displayed.
> 2. Remove all pinned queries and assert that there is no `Study Queries` panel at the top of the page.

Reviewer 2: ```suggestion
4. Assert that using the `Starred Only` checkbox filter only keeps the starred ones.
5. Assert that removing the `Starred Only` checkbox prints the same queries originally printed.

- Repeat 4. and 5. for `Shared Only`, `Named Only` and `No run times` checkboxes.

6. Assert that mixing checkboxes returns the right query selection. Only queries that match all of the checked conditions should be displayed.
7. Remove all pinned queries and assert that there is no `Study Queries` panel at the top of the page.
```

## Pin Tests

- Click the `Pin` icon to perform the following tests (ensure that you have 'admin' permission):

1. With the `query name` text fieldataqueryd empty, click the `Submit` button and assert that the error message reads:

`Must provide a query name to pin query as.`

- Enter a query name and uncheck the following:
- [ ] `Pin Study Query`
- [ ] `Pin Dashboard Summary`
- [ ] `Pin to Login Page`

> 2. Click 'Submit' and assert that the error message reads :

Reviewer 2: ```suggestion
2. Click 'Submit' and assert that the error message reads:
```

## Sync not sync

[ ] uncheck `Sync with selected fields` and add all visits from `Default Visits` box. Then click some fields in the list so that theyj are greyed out.

1. Assert that selected fields each have a visit selection box.

2. Assert that all visits displayed in the field lines are the same as those in the `Default Visits` box.

3. Assert that all visits are described in the selected fields in the right column.

Remove some visits from `Default Visits` box and click on `Other Fields`.

4. Assert that new field visits only reflects what the `Default Visits` box is showing.

5. Assert that the newly selected fields in the right column have the new default visits.

6. Check `Sync with selected field` checkbox and assert that all fields match the default.

7. Make sure the `Selected Fields` are still updateable by ex-ing out visits in the main table.

## Filters

Click on `Add Filters` in `Next Steps` and make sure no filter is already selected.

1. Assert that the sentence `Currently querying for ALL candidates` should be displayed under the light-blue notification.

> 2. Assert the light-blue notification saying `Note that only candidates which you have permission to access in LORIS are included in results. Number of results may vary from other users running the same query' is there`.

Reviewer 2: ```suggestion
2. Assert the light-blue notification saying `Note that only candidates which you have permission to access in LORIS are included in results. Number of results may vary from other users running the same query` is there.
```

3. Assert that a preview of the number of candidates matched is displayed in the top-right hand corner.

4. Make sure that the `Add Condition` button triggers field select modal.

Add a field as a filter

5. Assert that the condition now appears in the filter list.

Select your filter, and enter your condition, and click `Submit`. 

> 6. Assert that: `Add Condition` and `Import from CSV` buttons are now replaced by `Add "and" condition` and `Add "or" condition`.
> 
> 7. Assert that the top-right corner is updated.
> 
> 8. Remove the added conditions with the trashbin icon on the right.
> 
> 9. Assert that `Import from CSV` button features an upload modal.
> 
> 10. Assert that: sending something different than CSV ends with an `Invalid <Something>` alert.
> 
> 11. Test file import by creating and importing different files with the preset options (candidates vs. sessions, DCCID vs. PSCID, with or without headers).

Reviewer 2: Elsewhere in this doc the instructions are not numbered, but the assertions are. Removing the number here would be more consistent with the rest of it
```suggestion
6. Assert that: `Add Condition` and `Import from CSV` buttons are now replaced by `Add "and" condition` and `Add "or" condition`.

7. Assert that the top-right corner is updated.

Remove the added conditions with the trashbin icon on the right.

8. Assert that `Import from CSV` button features an upload modal.

9. Assert that: sending something different than CSV ends with an `Invalid <Something>` alert.

10. Test file import by creating and importing different files with the preset options (candidates vs. sessions, DCCID vs. PSCID, with or without headers).
```

Make several queries with various operators (AND/OR) and depths (condition groups)

> 12. Assert that these operators are working.

Reviewer 2: ```suggestion
11. Assert that these operators are working.
```
